File: easy_language/lexer.py
from .erreurs import ErreurLexicale

MOTS_CLES = {
    "soit", "affiche", "demande",
    "si", "alors", "sinon",
    "tantque", "pour", "de", "jusqua", "pas", "chaque", "dans",
    "fonction", "retourne",
    "vrai", "faux", "rien",
    "et", "ou", "non",
    "arrete", "continue",
    "importe", "comme",
}

SYMBOLES = [
    ("==", "EGAL_EGAL"), ("!=", "DIFFERENT"),
    ("<=", "INF_EGAL"), (">=", "SUP_EGAL"),
    ("+", "PLUS"), ("-", "MOINS"), ("*", "FOIS"), ("/", "DIVISE"), ("%", "MODULO"),
    ("=", "EGAL"), ("<", "INF"), (">", "SUP"),
    ("(", "PARO"), (")", "PARF"),
    ("[", "CROCHET_O"), ("]", "CROCHET_F"),
    (":", "DEUX_POINTS"), (".", "POINT"),
]


class Jeton:
    def __init__(self, type_, valeur, ligne):
        self.type = type_
        self.valeur = valeur
        self.ligne = ligne

    def __repr__(self):
        return f"Jeton({self.type}, {self.valeur!r})"


class Lexer:
    def __init__(self, source):
        self.lignes = source.split("\n")
        self.jetons = []
        self.pile_indentation = [0]

    def tokeniser(self):
        for numero, ligne_brute in enumerate(self.lignes, start=1):
            self._traiter_ligne(ligne_brute, numero)

        while len(self.pile_indentation) > 1:
            self.pile_indentation.pop()
            self.jetons.append(Jeton("DEDENT", None, len(self.lignes) + 1))

        self.jetons.append(Jeton("FIN", None, len(self.lignes) + 1))
        return self.jetons

    def _traiter_ligne(self, ligne_brute, numero):
        if "\t" in ligne_brute.split("#")[0]:
            raise ErreurLexicale(
                "les tabulations ne sont pas autorisees pour l'indentation, utilisez des espaces",
                numero,
            )

        sans_espaces = ligne_brute.lstrip(" ")
        indentation = len(ligne_brute) - len(sans_espaces)
        contenu = sans_espaces.split("#", 1)[0].rstrip()

        if contenu == "":
            return

        if indentation > self.pile_indentation[-1]:
            self.pile_indentation.append(indentation)
            self.jetons.append(Jeton("INDENT", indentation, numero))
        else:
            while indentation < self.pile_indentation[-1]:
                self.pile_indentation.pop()
                self.jetons.append(Jeton("DEDENT", None, numero))
            if indentation != self.pile_indentation[-1]:
                raise ErreurLexicale("indentation incoherente", numero)

        self._tokeniser_contenu(contenu, numero)
        self.jetons.append(Jeton("NOUVELLE_LIGNE", None, numero))

    def _tokeniser_contenu(self, contenu, numero):
        i = 0
        n = len(contenu)
        while i < n:
            c = contenu[i]

            if c == " ":
                i += 1
                continue

            if c == '"':
                j = i + 1
                valeur = []
                while j < n and contenu[j] != '"':
                    if contenu[j] == "\\" and j + 1 < n:
                        suivant = contenu[j + 1]
                        correspondance = {"n": "\n", "t": "\t", '"': '"', "\\": "\\"}
                        valeur.append(correspondance.get(suivant, suivant))
                        j += 2
                    else:
                        valeur.append(contenu[j])
                        j += 1
                if j >= n:
                    raise ErreurLexicale("chaine de texte non terminee", numero)
                self.jetons.append(Jeton("TEXTE", "".join(valeur), numero))
                i = j + 1
                continue

            if c.isdigit():
                j = i
                a_point = False
                while j < n and (contenu[j].isdigit() or (contenu[j] == "." and not a_point)):
                    if contenu[j] == ".":
                        a_point = True
                    j += 1
                brut = contenu[i:j]
                valeur = float(brut) if a_point else int(brut)
                self.jetons.append(Jeton("NOMBRE", valeur, numero))
                i = j
                continue

            if c.isalpha() or c == "_":
                j = i
                while j < n and (contenu[j].isalnum() or contenu[j] == "_"):
                    j += 1
                mot = contenu[i:j]
                if mot in MOTS_CLES:
                    self.jetons.append(Jeton(mot.upper(), mot, numero))
                else:
                    self.jetons.append(Jeton("IDENT", mot, numero))
                i = j
                continue

            trouve = False
            for symbole, nom in SYMBOLES:
                if contenu.startswith(symbole, i):
                    self.jetons.append(Jeton(nom, symbole, numero))
                    i += len(symbole)
                    trouve = True
                    break
            if trouve:
                continue

            raise ErreurLexicale(f"caractere inattendu: {c!r}", numero)

File: c/lexer.c
#include "lexer.h"

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct { const char *mot; TypeJeton type; } MotCle;

static const MotCle MOTS_CLES[] = {
    {"soit", T_SOIT}, {"affiche", T_AFFICHE}, {"demande", T_DEMANDE},
    {"si", T_SI}, {"alors", T_ALORS}, {"sinon", T_SINON},
    {"tantque", T_TANTQUE}, {"pour", T_POUR}, {"de", T_DE},
    {"jusqua", T_JUSQUA}, {"pas", T_PAS}, {"chaque", T_CHAQUE}, {"dans", T_DANS},
    {"fonction", T_FONCTION}, {"retourne", T_RETOURNE},
    {"vrai", T_VRAI}, {"faux", T_FAUX}, {"rien", T_RIEN},
    {"et", T_ET}, {"ou", T_OU}, {"non", T_NON},
    {"arrete", T_ARRETE}, {"continue", T_CONTINUE},
};
#define NB_MOTS_CLES (sizeof(MOTS_CLES) / sizeof(MOTS_CLES[0]))

static void erreur_lexicale(int ligne, const char *msg) {
    fprintf(stderr, "Erreur ligne %d: %s\n", ligne, msg);
    exit(1);
}

static void ajouter(ListeJetons *l, Jeton j) {
    if (l->compte >= l->capacite) {
        l->capacite = l->capacite ? l->capacite * 2 : 64;
        l->jetons = realloc(l->jetons, sizeof(Jeton) * l->capacite);
    }
    l->jetons[l->compte++] = j;
}

static Jeton jeton_simple(TypeJeton type, int ligne) {
    Jeton j = {0};
    j.type = type;
    j.ligne = ligne;
    return j;
}

typedef struct { const char *sym; TypeJeton type; } Symbole;
static const Symbole SYMBOLES[] = {
    {"==", T_EGAL_EGAL}, {"!=", T_DIFFERENT}, {"<=", T_INF_EGAL}, {">=", T_SUP_EGAL},
    {"+", T_PLUS}, {"-", T_MOINS}, {"*", T_FOIS}, {"/", T_DIVISE}, {"%", T_MODULO},
    {"=", T_EGAL}, {"<", T_INF}, {">", T_SUP},
    {"(", T_PARO}, {")", T_PARF}, {"[", T_CROCHET_O}, {"]", T_CROCHET_F},
    {":", T_DEUX_POINTS}, {",", T_VIRGULE},
};
#define NB_SYMBOLES (sizeof(SYMBOLES) / sizeof(SYMBOLES[0]))

static void tokeniser_contenu(ListeJetons *l, const char *contenu, int ligne) {
    int i = 0;
    int n = (int)strlen(contenu);
    while (i < n) {
        char c = contenu[i];
        if (c == ' ') { i++; continue; }

        if (c == '"') {
            int j = i + 1;
            char tampon[4096];
            int t = 0;
            while (j < n && contenu[j] != '"') {
                if (contenu[j] == '\\' && j + 1 < n) {
                    char suivant = contenu[j + 1];
                    char remplace = suivant;
                    if (suivant == 'n') remplace = '\n';
                    else if (suivant == 't') remplace = '\t';
                    tampon[t++] = remplace;
                    j += 2;
                } else {
                    tampon[t++] = contenu[j++];
                }
            }
            if (j >= n) erreur_lexicale(ligne, "chaine de texte non terminee");
            tampon[t] = '\0';
            Jeton jt = jeton_simple(T_TEXTE, ligne);
            jt.texte = strdup(tampon);
            ajouter(l, jt);
            i = j + 1;
            continue;
        }

        if (isdigit((unsigned char)c)) {
            int j = i;
            int a_point = 0;
            while (j < n && (isdigit((unsigned char)contenu[j]) || (contenu[j] == '.' && !a_point))) {
                if (contenu[j] == '.') a_point = 1;
                j++;
            }
            char brut[64];
            int longueur = j - i;
            memcpy(brut, contenu + i, longueur);
            brut[longueur] = '\0';
            Jeton jt;
            if (a_point) {
                jt = jeton_simple(T_NOMBRE_REEL, ligne);
                jt.reel = atof(brut);
            } else {
                jt = jeton_simple(T_NOMBRE_ENTIER, ligne);
                jt.entier = atoll(brut);
            }
            ajouter(l, jt);
            i = j;
            continue;
        }

        if (isalpha((unsigned char)c) || c == '_') {
            int j = i;
            while (j < n && (isalnum((unsigned char)contenu[j]) || contenu[j] == '_')) j++;
            int longueur = j - i;
            char mot[256];
            memcpy(mot, contenu + i, longueur);
            mot[longueur] = '\0';

            TypeJeton type_trouve = T_IDENT;
            for (size_t k = 0; k < NB_MOTS_CLES; k++) {
                if (strcmp(mot, MOTS_CLES[k].mot) == 0) { type_trouve = MOTS_CLES[k].type; break; }
            }
            Jeton jt = jeton_simple(type_trouve, ligne);
            if (type_trouve == T_IDENT) jt.texte = strdup(mot);
            ajouter(l, jt);
            i = j;
            continue;
        }

        int trouve = 0;
        for (size_t k = 0; k < NB_SYMBOLES; k++) {
            size_t sl = strlen(SYMBOLES[k].sym);
            if (strncmp(contenu + i, SYMBOLES[k].sym, sl) == 0) {
                ajouter(l, jeton_simple(SYMBOLES[k].type, ligne));
                i += (int)sl;
                trouve = 1;
                break;
            }
        }
        if (trouve) continue;

        char msg[64];
        snprintf(msg, sizeof(msg), "caractere inattendu: '%c'", c);
        erreur_lexicale(ligne, msg);
    }
}

ListeJetons lexer_tokeniser(const char *source) {
    ListeJetons resultat = {0};

    int pile_indent[256];
    int profondeur = 0;
    pile_indent[0] = 0;

    const char *curseur = source;
    int numero_ligne = 0;
    int continuer = 1;

    while (continuer) {
        numero_ligne++;

        const char *debut = curseur;
        while (*curseur != '\n' && *curseur != '\0') curseur++;
        int longueur = (int)(curseur - debut);
        if (longueur > 4000) longueur = 4000; /* protege le tampon fixe ci-dessous */
        char ligne_brute[4096];
        memcpy(ligne_brute, debut, longueur);
        ligne_brute[longueur] = '\0';

        if (*curseur == '\0') continuer = 0; else curseur++; /* saute le '\n' */

        /* enlever un eventuel '\r' final (fins de ligne Windows) */
        size_t len = strlen(ligne_brute);
        if (len > 0 && ligne_brute[len - 1] == '\r') ligne_brute[len - 1] = '\0';

        /* verifier tabulations avant le premier '#' */
        for (const char *p = ligne_brute; *p && *p != '#'; p++) {
            if (*p == '\t') erreur_lexicale(numero_ligne, "les tabulations ne sont pas autorisees pour l'indentation, utilisez des espaces");
        }

        int indentation = 0;
        while (ligne_brute[indentation] == ' ') indentation++;

        char contenu[4096];
        strncpy(contenu, ligne_brute + indentation, sizeof(contenu) - 1);
        contenu[sizeof(contenu) - 1] = '\0';
        char *croisillon = strchr(contenu, '#');
        if (croisillon) *croisillon = '\0';
        int fin = (int)strlen(contenu);
        while (fin > 0 && contenu[fin - 1] == ' ') { contenu[--fin] = '\0'; }

        if (contenu[0] == '\0') {
            continue;
        }

        if (indentation > pile_indent[profondeur]) {
            profondeur++;
            pile_indent[profondeur] = indentation;
            ajouter(&resultat, jeton_simple(T_INDENT, numero_ligne));
        } else {
            while (indentation < pile_indent[profondeur]) {
                profondeur--;
                ajouter(&resultat, jeton_simple(T_DEDENT, numero_ligne));
            }
            if (indentation != pile_indent[profondeur]) {
                erreur_lexicale(numero_ligne, "indentation incoherente");
            }
        }

        tokeniser_contenu(&resultat, contenu, numero_ligne);
        ajouter(&resultat, jeton_simple(T_NOUVELLE_LIGNE, numero_ligne));
    }

    while (profondeur > 0) {
        profondeur--;
        ajouter(&resultat, jeton_simple(T_DEDENT, numero_ligne + 1));
    }
    ajouter(&resultat, jeton_simple(T_FIN, numero_ligne + 1));

    return resultat;
}
